pointcloud/basic_octree.go

```diff
@@ -129,3 +131,55 @@ func (octree *BasicOctree) Iterate(numBatches, currentBatch int, fn func(p r3.Ve
 func (octree *BasicOctree) MetaData() MetaData {
 	return octree.meta
 }
+
+// CollidesWithGeometry will return whether a given geometry is in collision with a given point.
+func (octree *BasicOctree) CollidesWithGeometry(geom spatialmath.Geometry, threshold int, buffer float64) (bool, error) {
+	if octree.MaxVal() < threshold {
+		return false, nil
+	}
+	ocbox, err := spatialmath.NewBox(
+		spatialmath.NewPoseFromPoint(octree.center),
+		r3.Vector{octree.sideLength + buffer, octree.sideLength + buffer, octree.sideLength + buffer},
+		"",
+	)
+	if err != nil {
+		return false, err
+	}
+
+	// Check whether our geom collides with the area represented by the octree. If false,
+	collide, err := geom.CollidesWith(ocbox)
+	if err != nil {
+		return false, err
+	}
+	if !collide {
+		return false, nil
+	}
+
+	if octree.node.point != nil {
+		ptGeom, err := spatialmath.NewSphere(spatialmath.NewPoseFromPoint(octree.node.point.P), buffer, "")
+		if err != nil {
+			return false, err
+		}
+
+		ptCollide, err := geom.CollidesWith(ptGeom)
+		if err != nil {
+			return false, err
+		}
+		if ptCollide {
+			return true, nil
+		}
+	}
+
+	if len(octree.node.children) > 0 {
```

Comment:
[major] Stick with the current octree pattern of checking node types.

This case is the internal node case, emptyLeafNode should always return false and filledLeafNode should have the checks from line 137-172

Reply:
That would allow the reduction of some collision checks and be more efficient, good catch.